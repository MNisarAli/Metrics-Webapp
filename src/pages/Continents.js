import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { getContinents } from '../redux/Continent/continentsSlice';
import Dashboard from '../components/Dashboard';
import Wrapper from '../components/Wrapper';

const Continents = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const { continents, status } = useSelector((state) => state.continents);

  useEffect(() => {
    // Fetch continents data from API only when it is not already available.
    if (continents.length === 0) {
      dispatch(getContinents());
    }
  }, [continents, dispatch]);

  // Show loading message when data is being fetched.
  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  // Calculate total population of all continents.
  const totalPopulation = continents.reduce((acc, obj) => acc + obj.population, 0);

  const worldMap = 'https://svgsilh.com/svg/306338.svg';

  // Handle click event when a continent is clicked
  const handleContinentClick = (continent) => {
    navigate(`/continent?continent=${JSON.stringify(continent)}`);
  };

  return (
    <>
      {/* Show dashboard with world map and total population */}
      <Dashboard map={worldMap} name="World" population={totalPopulation} />

      <section className="data-section">
        <h3 className="section-title">Continents</h3>
        <nav className="navbar">
          {/* Iterate over continents array and display each continent using Wrapper component */}
          {continents.map((continent) => (
            <a
              className="nav-item"
              key={continent.id}
              href={continent.path}
              onClick={(e) => {
                e.preventDefault();
                handleContinentClick(continent);
              }}
            >
              <Wrapper wrapper={continent} />
            </a>
          ))}
        </nav>
      </section>
    </>
  );
};

export default Continents;
