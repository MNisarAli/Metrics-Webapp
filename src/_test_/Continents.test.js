import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import Continents from '../pages/Continents';
import store from '../Redux/configureStore';

describe('Continents page', () => {
  test('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Continents />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
