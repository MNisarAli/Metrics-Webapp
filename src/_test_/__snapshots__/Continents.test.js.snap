// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Continents page renders correctly 1`] = `
Array [
  <section
    className="dashboard"
  >
    <img
      alt="World"
      className="dashboard-map"
      src="https://svgsilh.com/svg/306338.svg"
    />
    <article
      className="title-container"
    >
      <h2
        className="page-title"
      >
        World
      </h2>
      <span>
        0
      </span>
    </article>
  </section>,
  <section
    className="data-section"
  >
    <h3
      className="section-title"
    >
      Continents
    </h3>
    <nav
      className="navbar"
    />
  </section>,
]
`;
